pkg/cli/gettoken/oidc/client.go

```diff
@@ -0,0 +1,229 @@
+package oidc
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	gooidc "github.com/coreos/go-oidc/v3/oidc"
+	"github.com/int128/oauth2cli"
+	"golang.org/x/oauth2"
+
+	"github.com/openshift/oc/pkg/cli/gettoken/pkce"
+)
+
+// Authenticator defines the basic functionality to support Auth Code FLow
+type Authenticator interface {
+	GetTokenByAuthCode(ctx context.Context, input GetTokenByAuthCodeInput, localServerReadyChan chan<- string) (string, string, error)
+	Refresh(ctx context.Context, refreshToken string) (string, string, error)
+	SupportedPKCEMethods() []string
+}
+
+// ExchangeAuthCodeInput uses the fields to exchange auth code and tokens
+type ExchangeAuthCodeInput struct {
+	Code        string
+	PKCEParams  pkce.Params
+	Nonce       string
+	RedirectURI string
+}
+
+// GetTokenByAuthCodeInput stores the fields to get auth code to be used in exchange process.
+type GetTokenByAuthCodeInput struct {
+	BindAddress            []string
+	State                  string
+	Nonce                  string
+	PKCEParams             pkce.Params
+	RedirectURLHostname    string
+	AuthRequestExtraParams map[string]string
+	LocalServerSuccessHTML string
+	LocalServerCertFile    string
+	LocalServerKeyFile     string
+}
+
+type client struct {
+	httpClient           *http.Client
+	provider             *gooidc.Provider
+	oauth2Config         oauth2.Config
+	supportedPKCEMethods []string
+}
+
+func (c *client) GetTokenByAuthCode(ctx context.Context, input GetTokenByAuthCodeInput, localServerReadyChan chan<- string) (string, string, error) {
```

Comment:
It should be reference type not only because of similarity with https://github.com/int128/kubelogin/blob/79a6f8596774d9d1fff2319ddc84a3a87f80fa48/pkg/oidc/client/client.go#L75, I think more importantly, to pass `httpClient` as reference

Reply:
I meant the "input" struct. I don't think we need to create its shallow copy on function call, do we?